crates/emitter/src/forward_jump_emitter.rs

```diff
@@ -0,0 +1,73 @@
+use super::emitter::{BytecodeOffset, EmitError};
+use crate::ast_emitter::AstEmitter;
+
+#[derive(Debug)]
+pub enum JumpKind {
+    Coalesce,
+    LogicalAnd,
+    LogicalOr,
+    IfEq,
+    Goto,
+}
+
+#[derive(Debug)]
+#[must_use]
+pub struct JumpPatchEmitter {
+    pub offset: BytecodeOffset,
+}
+impl JumpPatchEmitter {
+    pub fn emit(self, emitter: &mut AstEmitter) -> Result<(), EmitError> {
+        emitter.emit.patch_jump_target(vec![self.offset]);
+        emitter.emit.jump_target();
+        Ok(())
+    }
+}
+
+impl Drop for JumpPatchEmitter {
+    fn drop(&mut self) {
+        panic!("Dropped JumpPatchEmitter without calling it");
```

Comment:
what happens if the caller fails in between `ForwardJumpEmitter.emit` and `JumpPatchEmitter.emit` ?
(for example, by hitting `NotImplemented`)
does it panic?

Reply:
yes, it will panic... is there another way to make sure we use jumpPatchEmitter if it is there? i noticed that must_use doesn't work if there is an if branch:

```
let a = ForwardJumpEmitter { ... }.emit(self)?;

if (x) {
  a.patch(self)?;
} else {
  // forget to patch
}
```

Reply:
the way https://github.com/mozilla-spidermonkey/jsparagus/pull/257 took is that,
let the helper handle everything.

So, the consumer passes closures for code between control structure,
and helper emits all opcodes about control structure.
So that, as long as the helper correctly works, the consumer is safe.
(the code inside helper is unsafe, but that's what helper is for)

but I think that's not directly applicable to this helper struct,
because multiple jumps can overlap, like if-else case.

If we add helper for if-else with closure-passing style,
and let it handle everything about jump by using these structs,
the consumer of the if-else helper should be safe.
(again, as long as if-else helper is correct)

So, how about the following:
  * `AstEmitter` shouldn't directly touch `ForwardJumpEmitter`/`JumpPatchEmitter`
  * Add helper structs (with closure-passing style) that uses `ForwardJumpEmitter`/`JumpPatchEmitter`
in that way, we can reduce the possibility of the bug (that caused by forgetting patching, or wrongly patching),
without adding `Drop` trait to `JumpPatchEmitter`.

For this PR, I think we can just remove the `Drop` part,
and leave those new helpers to other issues.